### PR TITLE
route: get all per filter config directly

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: route: get all per filter config directly
Additional Description:

1. used absl::string_view rather than `const std::string&` as input parameter type. This allow the `mostSpecificPerFilterConfig()` and `perFilterConfigs()` to take a string view as input parameter. 
2. replace the `traversePerFilterConfig()` with `perFilterConfigs()`. The `perFilterConfigs()` will return a vector (inlined) of configs directly and needn't the caller to provide a callback function to collect them. The new method is more straight forward and easier to use. The new method didn't effect any exist logic, but reduce ~150 lines code.

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Review by ravenblackx — APPROVED** [review]

File: envoy/router/router.h [hunk 1]
```diff
@@ -122,6 +122,7 @@ class RouteSpecificFilterConfig {
   virtual ~RouteSpecificFilterConfig() = default;
 };
 using RouteSpecificFilterConfigConstSharedPtr = std::shared_ptr<const RouteSpecificFilterConfig>;
+using RouteSpecificFilterConfigs = absl::InlinedVector<const RouteSpecificFilterConfig*, 4>;
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply envoy/router/router.h:125]: Then we need to wraper it with a reference wrapper which actually is not very convenient, considering in most case, we will do a dynamic_cast after this calling anyway.

[the remark this answers is not in the document]
Comment (alyssawilk) [reply envoy/router/router.h:125]: I thought we were doing continue so not adding to the vector for any instances which were null?  am I mixing up my functions?

[the remark this answers is not in the document]
Comment (wbpcode) [reply envoy/router/router.h:125]: You are basically right. But note this is a vector of base type pointers which don't provide any methods to access configuration.

If a filter want to access the configuration, it still needs to do a dynamic_cast to these pointers.  And the dynamic_cast may still return a nullptr.

So, return a pointer here is more convenient.

```
for (const auto* cfg : ...) {
  const auto* typed_cfg = dynamic_cast<const TypedConfig*>(cfg);
}

// If reference wrapper is used (a little minor complex. And some fresh developers may try to dynamic_cast a reference directly incorrectly, which may throw runtime exception. So, I think pointer is better here.)

for (const auto& cfg : ...) {
  const auto* typed_cfg = dynamic_cast<const TypedConfig*>(&cfg.get());
}


```

**Pull request merged**